File: JAVA/Triangle.java
public class Solution {
    public int minimumTotal(ArrayList<ArrayList<Integer>> triangle) {
        int len = triangle.size();
        if (len == 0) return 0;
        int[] sum = new int[len];
        sum[0] = triangle.get(0).get(0);
        for (int i = 1; i < len; i++) {
            ArrayList<Integer> arr = triangle.get(i);
            int tmp = sum[0];
            for (int j = 0; j < i+1; j++) {
                int min = arr.get(j) + tmp;
                if (j + 1 < i) { tmp = Math.min(sum[j], sum[j+1]); }
                else { tmp = sum[j]; }
                sum[j] = min;
            }   
        }
        int min = sum[0];
        for (int i = 0; i < len; i++) {
            if (sum[i] < min) min = sum[i];
        }
        return min;
    }
}

File: Triangle.cpp
class Solution {
public:
    int minimumTotal(vector<vector<int> > &triangle) {
        if (triangle.size() == 0) return 0;
        int m = triangle.size();
        vector<int> dp(m);
        for (int i = 0; i < m; i++) {
            vector<int> tmp(i + 1);
            for (int j = 0; j <= i; j++) {
                if (i == 0) {
                    tmp[0] = triangle[0][0];
                } else if (j == 0) {
                    tmp[0] = dp[0] + triangle[i][0];
                } else if (j == i) {
                    tmp[j] = dp[j - 1] + triangle[i][i];
                } else {
                    tmp[j] = min(dp[j - 1], dp[j]) + triangle[i][j];
                }
            }
            for (int j = 0; j <= i; j++) {
                dp[j] = tmp[j];
            }
        }
        int result = dp[0];
        for (int i = 0; i < dp.size(); i++) {
            result = min(result, dp[i]);
        }
        return result;
    }
};
